lib/puppet/functions/lookup.rb

```diff
@@ -0,0 +1,199 @@
+# Looks up data defined using Hiera and Data Providers
+# The function is callable with one to four arguments and optionally with a code block to provide default value.
+#
+# The lookup function can be called in one of these ways:
+#
+#     lookup(name)
+#     lookup(name, value_type)
+#     lookup(name, value_type, default_value)
+#     lookup(name, value_type, default_value, merge)
+#     lookup(options_hash)
+#     lookup(name, options_hash)
+#
+# The function may optionally be called with a code block / lambda with the following signature:
+#
+#     lookup(...) |$name| { ... }
+#
+# The block, if present, is mutually exclusive to the `default_value` and will be called with the `name` used in
+# in the lookup when no value is found. The value produced by the block then becomes the value produced by the
+# lookup.
+#
+# When a block is used, it is the users responsibility to call `error` if an undef value is not acceptable.
+#
+# The content of the options hash is:
+#
+# * `name` - The name or array of names to lookup (first found is returned)
+# * `value_type` - The type to assert (a Type or a type specification in string form)
+# * `default_value` - The default value if there was no value found (must comply with the data type)
+# * `accept_undef` - (default `false`) An `undef` result is accepted if this options is set to `true`.
+# * `override` - a hash with map from names to values that are used instead of the underlying bindings. If the name
+#   is found here it wins. Defaults to an empty hash.
+# * `extra` - a hash with map from names to values that are used as a last resort to obtain a value. Defaults to an
```

Comment:
The name "extra" (while used in hiera) is a really bad name, it is really `default_data` - suggest changing this name.